File: scala/rna-transcription/src/main/scala/Dna.scala
//import scalaz.std.list._
//import scalaz.std.option._
//import scalaz.syntax.traverse._
import cats.instances.list._
import cats.instances.option._
import cats.syntax.traverse._


object Dna {
  private type Strand = String
  private type Nucleotide = Char

  def toRna(dna: String): Option[Strand] =
    dna.toList.traverse(toRnaNucleotide) map (_.mkString)

  private val DnaToRna =
    Map('G' -> 'C', 'C' -> 'G', 'T' -> 'A', 'A' -> 'U')

  private def toRnaNucleotide(char: Char): Option[Nucleotide] =
     DnaToRna get char
}

File: haskell/rna-transcription/DNA.hs
module DNA (toRNA) where

import qualified Data.Map as Map
import Data.Traversable

toRnaNucleotide :: Map.Map Char Char
toRnaNucleotide = Map.fromList [ ('G','C'), ('C','G'), ('T','A'), ('A','U') ]

toRNA :: (Traversable t) => t Char -> Maybe (t Char)
toRNA = traverse (`Map.lookup` toRnaNucleotide)
